tests/unit/s2n_client_alpn_extension_test.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+
+#include "tls/extensions/s2n_client_alpn.h"
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    const char *protocols[] = { "protocol1", "protocol2", "protocol2" };
+    const uint8_t protocols_count = s2n_array_len(protocols);
+
+    struct s2n_connection *conn;
+    EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+    EXPECT_NULL(s2n_get_application_protocol(conn));
```

Comment:
This function gets the value of `conn->application_protocol` and doesn't check for any overrides, shouldn't it check for `conn->application_protocol_overriden` value as well? 

Looks like we have `application_protocol` and `application_protocol_overridden` in s2n_connection and `application_protocol` in s2n_config. Do we need `application_protocol` set both in config and connection?

Reply:
These variables have kind of confusing names. "application_protocol**s**" (plural) and "application_protocols_overriden" are actually **preferences** for application protocols. The user sets those. "application_protocol" (singular) is the protocol agreed on in the handshake, after comparing the preferences sent by the client to the server's preferences.